src/main/scala/org/ensime/core/Completion.scala

```diff
@@ -110,132 +114,92 @@ trait CompletionControl {
-          typeSearchSyms)
-      case Some(SymbolContext(scPosition, scPrefix, constructing)) =>
-        askReloadFile(scPosition.source)
-        val x = new Response[List[Member]]
-        askScopeCompletion(scPosition, x)
-        (scPrefix, makeAll(x, scPrefix, constructing))
-      case Some(MemberContext(mcPosition, mcPrefix, constructing)) =>
-        askReloadFile(mcPosition.source)
-        val x = new Response[List[Member]]
-        askTypeCompletion(mcPosition, x)
-        (mcPrefix, makeAll(x, mcPrefix, constructing))
-      case _ =>
-        logger.error("Unrecognized completion context.")
-        ("", List.empty)
-    }
-    CompletionInfoList(prefix, results.sortWith({ (c1, c2) =>
-      c1.relevance > c2.relevance ||
-        (c1.relevance == c2.relevance &&
-          c1.name.length < c2.name.length)
-    }).take(maxResults))
-  }
-
-  private val ident = "[a-zA-Z0-9_]"
-  private val nonIdent = "[^a-zA-Z0-9_]"
-  private val ws = "[ \n\r\t]"
-
-  trait CompletionContext
+    val preceding = inputP.source.content.slice(
+      Math.max(0, inputP.point - 100), inputP.point)
 
-  private val packRE = ("^.*?(?:package|import)[ ]+((?:[a-z0-9]+\\.)*)(?:(" + ident + "*)|\\{.*?(" + ident + "*))$").r
-  case class PackageContext(path: String, prefix: String) extends CompletionContext
-  def packageContext(preceding: String): Option[PackageContext] = {
-    if (packRE.findFirstMatchIn(preceding).isDefined) {
-      val m = packRE.findFirstMatchIn(preceding).get
-      if (m.group(2) != null) {
-        logger.info("Matched package context: " + m.group(1) + "," + m.group(2))
-        Some(PackageContext(m.group(1), m.group(2)))
-      } else {
-        logger.info("Matched package context: " + m.group(1) + "," + m.group(3))
-        Some(PackageContext(m.group(1), m.group(3)))
-      }
-    } else None
-  }
-
-  private val nameFollowingWhiteSpaceRE =
-    List("^", ws, "*", "(", ident, "*)$").mkString.r
-
-  private val nameFollowingReservedRE =
-    List("(?:", nonIdent, "|", ws, ")",
-      "(?:else|case|new|with|extends|yield|return|throw|do)",
-      ws, "+",
-      "(", ident, "*)$").mkString.r
-
-  private val nameFollowingSyntaxRE =
-    List("([!:=><\\(,;\\}\\[\\{\n+*/\\^&~%\\-])",
-      ws, "*", "(", ident, "*)$").mkString.r
-
-  private val constructorNameRE =
-    List("(?:", nonIdent, "|", ws, ")", "(?:new)", ws, "+",
-      "(", ident, "*)$").mkString.r
-
-  private val nameFollowingControl =
-    List("(?:for|while|if)", ws, "*", "\\((.*?)\\)", ws, "*",
-      "(", ident, "*)$").mkString.r
-
-  // Return true if the given string is balanced with respect
-  // to parenthesis. Ignores the possibility of comments and strings.
-  private def parenBalanced(s: String): Boolean = {
-    var i = 0
-    var p = 0
-    while (i < s.length) {
-      val ch = s.charAt(i)
-      if (ch == '(') p += 1
-      else if (ch == ')') p -= 1
-      i += 1
+    val defaultPrefix = IdentRegexp.findFirstMatchIn(preceding) match {
```

Comment:
I accept I may be showing complete failure to understand the code, but surely you want the lastMatchIn - you are a a point P and you want the nearest Ident backwards, rather than the first match from 100 characters previously?